crates/wasmparser/src/binary_reader.rs

```diff
@@ -1553,6 +1553,726 @@ impl<'a> BinaryReader<'a> {
         Ok(BlockType::FuncType(idx as u32))
     }
 
+    /// Reads the next available `Operator` and calls the respective visit method.
+    ///
+    /// # Errors
+    ///
+    /// If `BinaryReader` has less bytes remaining than required to parse
+    /// the `Operator`.
+    #[rustfmt::skip]
+    pub fn visit_operator<T>(&mut self, visitor: &mut T) -> Result<<T as VisitOperator<usize>>::Output>
```

Comment:
I actually have a personal use case where I would not like to return a `Result` for any of the visitation methods. One idea is to have a `where E: From<BinaryReadyError>` bound in order to flatten the `Result`. However, the current design is more flexible since it allows the user to take control over this behavior. The user can still perform this flattening. In benchmarks I tried to flatten but it actually was bad for the performance so I reverted it.

Reply:
The base layer is always returning a `Result` though because an operator may not be encoded correctly. Once there's one layer of a `Result` I would assume that each individual method returning a `Result` would be trivially inline-able by LLVM to statically determine whether that arm never returned an error or not.

I could try to help look into regressing benchmarks though if you've got those.

Reply:
In the common `wasmparser` benchmarks I saw the following differences:

- Current Method: roughly 10% improvements compared to `master`
- Flattening `Result`: just roughly 5% improvements compared to `master`

I expected both versions to have equal performance but this measured slow-down kinda defeats the purpose of this optimization. The experimental flattening was implemented on the `VisitOperator` impl for `FuncValidator`.

---

Could we maybe experiment with flattening of `Result`s in a follow-up PR? I feel like this PR is already kinda pumped and would like to chunk up work into smaller pieces.